src/aero_kernel/src/syscall/process.rs

```diff
@@ -310,6 +311,80 @@ pub fn sigaction(
     Ok(0)
 }
 
+pub fn ipc_send(
+    pid: usize,
+    buf_ptr: usize,
+    buf_len: usize,
+    tag: usize,
+    flags: usize,
+) -> Result<usize, AeroSyscallError> {
+    let scheduler = scheduler::get_scheduler();
+    let current_task = scheduler.current_task();
+    let target_task = scheduler
+        .find_task(TaskId::new(pid))
+        .ok_or(AeroSyscallError::EINVAL)?;
+
+    let _flags = IpcSendFlags::from_bits_truncate(flags as u32);
+    let buf = unsafe { core::slice::from_raw_parts(buf_ptr as *const u8, buf_len) };
+
+    let queue = target_task.message_queue();
+    let msg_id = queue.make_id();
+
+    queue.push_back(Message::new(msg_id, tag, current_task.pid(), buf));
+    scheduler.wake_up(target_task);
+
+    Ok(msg_id)
+}
+
+pub fn ipc_recv(
+    buf_ptr: usize,
+    buf_len: usize,
+    pid_ptr: usize,
+    length_ptr: usize,
+    tag_ptr: usize,
+    flags: usize,
+) -> Result<usize, AeroSyscallError> {
+    let scheduler = scheduler::get_scheduler();
+    let current_task = scheduler.current_task();
+    let queue = current_task.message_queue();
+
+    let buffer = unsafe { core::slice::from_raw_parts_mut(buf_ptr as *mut u8, buf_len) };
+    let pid = unsafe { &mut *(pid_ptr as *mut usize) };
+    let length = unsafe { &mut *(length_ptr as *mut usize) };
+    let tag = unsafe { &mut *(tag_ptr as *mut usize) };
+    let flags = IpcRecvFlags::from_bits_truncate(flags as u32);
+
+    if queue.is_empty() && flags.contains(IpcRecvFlags::NO_WAIT) {
+        return Err(AeroSyscallError::ENOMSG);
+    }
+
+    let message = if queue.is_empty() {
+        queue.wait_for_message()?
+    } else {
+        queue.pop_front().unwrap()
```

Comment:
Instaed of popping it first and if the buffet cannot fit it, maybe to peek the next item and if it satisfies the requirements you can pop it.

Reply:
Yeah, I thought of that too but I couldn’t figure out a clean way to do it because MessageQueue internally uses a Mutex so I can’t just borrow the message, any ideas?

Reply:
Fixed in c2f9e56677a965209df97c88405c56180b22e52d